### ISSUE TITLE
$refs in parameter is not processed

### ISSUE THREAD

**Issue by ioggstream** [koxudaxi/fastapi-code-generator#24]
## I expect

$refs in parameters to be processed, eg

```
paths:
  /foo:
    parameters:
    - $ref: '#/components/parameters/MyParam'
components:
  parameters:
    MyParam:
      name: foo
      schema:
        type: string
```

## Instead

I get KeyError

```
  File "/usr/local/lib/python3.8/site-packages/fastapi_code_generator/parser.py", line 207, in get_parameter_type
    schema: JsonSchemaObject = JsonSchemaObject.parse_obj(parameter["schema"])
KeyError: 'schema'
```

**Comment by koxudaxi** [comment]
@ioggstream 
Thank you for creating this issue.
We should support `$ref` in parameters.

**Comment by ioggstream** [comment]
A solution could be to resolve all the $refs and then run the vanilla parser, which will then be `ref` - free.

WDYT?

**Comment by koxudaxi** [comment]
@ioggstream 
I think it's good :nerd_face:

**Comment by slym** [comment]
I'm facing the same issue with latest version. Is it supposed to be fixed or are you still working on it ?

**Comment by koxudaxi** [comment]
@slym 
Yes, I'm working on it.
I must resolve some problems.
1. datamodel-code-generator don't parse `/components/parameters` (fastapi-code-generator use datamodel-code-generator to generate models)
2. fastapi-code-generator can't get parse name from datamodel-code-generator. (duplicated models will be renamed)

I will update datamodel-code-generator for compatibility :)

**Comment by koxudaxi** [comment]
@ioggstream @slym 
I have released  a new version as `0.0.16`

This version doesn't parse complex models perfectly. But, I guess covering many cases

**Comment by slym** [comment]
I confirm it is now working well with version 0.0.16 even with some nested refs. Great work ! Thanks.

**Comment by ioggstream** [comment]
Thanks! As it still does not support remote refs, I'll close this and file another! Great work!

**Comment by oktavlachs** [comment]
Hey @koxudaxi @ioggstream,
thank you so much for your work on this project, it really helps a lot!
Today I ran into this issue with a url ref and thought it might be a bug within the scope of this thread:

```
paths:
  /{supiOrSuci}/security-information-rg:
    get:
      summary: Get authentication data for the FN-RG
      operationId: GetRgAuthData
      tags:
        - Get Auth Data for FN-RG
      parameters:
        - name: plmn-id
          in: query
          description: serving PLMN ID
          content:
            application/json:
              schema:
                $ref: 'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/PlmnId'
```

Error was:
```
site-packages/fastapi_code_generator/parser.py", line 293, in get_parameter_type
    schema: JsonSchemaObject = JsonSchemaObject.parse_obj(parameter["schema"])
KeyError: 'schema'
```
Should your gernerator also cover this nested content type?
Thanks!

**Comment by koxudaxi** [comment]
@oktavlachs 
Thank you for posting the problem.
I have not known that `parameters` can accept `content`.
OK, I will fix it.

**Comment by ioggstream** [comment]
@koxudaxi not sure that `parameters` can accept `content` in OAS3 though.

**Comment by koxudaxi** [comment]
@ioggstream 
Yes, Me too :sweat_smile:

**Comment by oktavlachs** [comment]
I work with a quite big, complex and nested 3gpp OAS3 in this case ;)

**Comment by oktavlachs** [comment]
@koxudaxi @ioggstream 
Could you give me an update how you will proceed?
Additionally: Will you reopen this issue or should I create a new one?
Thanks :)

**Comment by koxudaxi** [comment]
@oktavlachs 
I'm sorry for the late reply.

I have released a new version as `0.2.5`
Would please check it?

In this week, I'm trying to support generic type in pydantic-pycharm-plugin.
https://github.com/koxudaxi/pydantic-pycharm-plugin/issues/275
I will update fastapi-code-generator after the work.

**Comment by oktavlachs** [comment]
Again, thanks a lot for taking care of this issue so quickly! :1st_place_medal: 

I am one step further, but now I am facing an issue in the "datamodel-code-generator" dependency.
But I am unsure whether this is an issue there, an issue as a result of your fix, or an issue at all.
Maybe you have a clue?

It seems that a lot of classes in the specs are not resolved. Neither local spec references (#components/...), which are litereally all defined within the same file. Nor URL references (http://localhost:8081...). I cannot pinpoint the error here as the error description is misleading.

```
Traceback (most recent call last):
  File "xxx/venv/bin/fastapi-codegen", line 8, in <module>
    sys.exit(app())
  File "xxx/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "xxx/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "xxx/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "xxx/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "xxx/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "xxx/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/xxx/fastapi_code_generator/__main__.py", line 26, in main
    return generate_code(input_name, input_text, output_dir, template_dir)
  File "xxx/fastapi_code_generator/__main__.py", line 72, in generate_code
    results = model_parser.parse()
  File "xxx/datamodel_code_generator/parser/base.py", line 407, in parse
    _, sorted_data_models, require_update_action_models = sort_data_models(
  File "xxx/datamodel_code_generator/parser/base.py", line 107, in sort_data_models
    return sort_data_models(
  File "xxx/datamodel_code_generator/parser/base.py", line 107, in sort_data_models
    return sort_data_models(
  File "xxx/datamodel_code_generator/parser/base.py", line 107, in sort_data_models
    return sort_data_models(
  [Previous line repeated 97 more times]
  File "xxx/datamodel_code_generator/parser/base.py", line 157, in sort_data_models
    raise Exception(f'A Parser can not resolve classes: {unresolved_classes}.')

Exception: A Parser can not resolve classes: [class: #/components/schemas/SubscriptionDataSets references: 
frozenset({'#/components/schemas/LcsMoData', '#/components/schemas/UeContextInSmsfData', 'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/TraceData', '#/components/schemas/SmsSubscriptionData', '#/components/schemas/V2xSubscriptionData', '#/components/schemas/SessionManagementSubscriptionData', '#/components/schemas/LcsPrivacyData', '#/components/schemas/AccessAndMobilitySubscriptionData', '#/components/schemas/UeContextInSmfData', '#/components/schemas/SmfSelectionSubscriptionData', '#/components/schemas/UeContextInAmfData', '#/components/schemas/SmsManagementSubscriptionData', '#/components/schemas/LcsBroadcastAssistanceTypesData'})], [class: http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/TraceData references: frozenset({'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/Ipv6Addr', 'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/Ipv4Addr', 'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/TraceDepth'})], [class: #/components/schemas/UeContextInSmsfData references: frozenset({'#/components/schemas/SmsfInfo'})], [class: #/components/schemas/SmsfInfo references: frozenset({'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/PlmnId', 'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/NfInstanceId'})], ...
```

There are a lot more lines with unresolved references but you get the idea.
Thanks!

**Comment by koxudaxi** [comment]
@oktavlachs 
I guess `http://localhost:8081/...` is not resolved. :thinking: 
The error will be shown when any unresolved $ref is found.

Would you please upload the schema?
I want to check the schema to investigate the error.

**Comment by oktavlachs** [comment]
@koxudaxi
I am pretty sure that the url is resolvable. I let wireshark run while executing the generator and none of the HTTP requests returned a 404.

The schemas are quite nested though with additionals $ref.

As an example: I found something in the logs for unresolved references which looks different from the others. Is this a valid reference within your library?

**Schema**
```
components:
  schemas:
    Area:
      type: object
      oneOf:
        - required:
          - tacs
        - required:
          - areaCode
      properties:
        tacs:
          type: array
          items:
            $ref: '#/components/schemas/Tac'
          minItems: 1
        areaCode:
            $ref: '#/components/schemas/AreaCode'
```

**Error log**
```
 [class: http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/Area references: frozenset({'http://localhost:8081/29.571/16.7.0/TS29571_CommonData.yaml#/components/schemas/Area/Area#-datamodel-code-generator-#-object-#-special-#'})]
```

**Comment by koxudaxi** [comment]
@oktavlachs 

> I am pretty sure that the url is resolvable. I let wireshark run while executing the generator and none of the HTTP requests returned a 404.

Great!! :)

```yaml
        tacs:
          type: array
          items:
            $ref: '#/components/schemas/Tac'
          minItems: 1
        areaCode:
            $ref: '#/components/schemas/AreaCode'

```
Good catch!!
The $ref expect case-sensitive.
Can you change the $ref to an actual path?

> The full schema section is over 2600 lines of code, ;) But I can provide it if that helps more. Let me know.

If you can't resolve the problem that way, then  I want to check all lines.

**Comment by oktavlachs** [comment]
@koxudaxi Sorry for taking so long.
I wasn't able to solve the issue, I am afraid.

As you suggested, here is the link to the yaml files I use to generate the code with:

https://github.com/jdegre/5GC_APIs

More specific, I am trying to generate this server https://github.com/jdegre/5GC_APIs/blob/master/TS29503_Nudm_SDM.yaml

I replaced the $ref to other files with a URL, which pointed to a simple static web server on localhost. That worked pretty well, but the code wouldn't generate with issues like the one described above.

Because these are official 3gpp specs, I would assume in general that it should also work with your library. Would you agree?
Thanks for your help and let me know if I can somehow provide more information. Thanks! :)

**Comment by koxudaxi** [comment]
@oktavlachs 
Thank you for giving the source.

Unfortunately, the fastapi-code-generator's model parser is not complete.
But, I have a plan.
I'm working on implementing an OpenAPI parser in the datacode-model-generator.
https://github.com/koxudaxi/datamodel-code-generator/pull/438

after I finish it, I will change that the fastapi-code-generator uses the parser for generating parameter models :wink:

**Comment by oktavlachs** [comment]
@koxudaxi 

That sounds great! 🙂
Do you roughly foresee how long that will take? Is it a small or a big task?

**Comment by koxudaxi** [comment]
> Do you roughly foresee how long that will take? Is it a small or a big task?

It's a big task!! 
I have a road map.
1. add OpenAPI path parser to datamodel-code-generator
2. add extension interface on datamodel-code-generator for a third-pary library like fastapi-code-generator
3. refactor fastapi-code-generator for the extension interface.
5. check your schema with the new version of fastapi-code-generator.

I would complete step 1 this week. 

Btw, I'm busy recently. I want to get more time to work on it. 😢

**Comment by oktavlachs** [comment]
@koxudaxi 
Sounds like a plan!

I also just discovered that there has been a FastAPI generator added to the popular OpenAPITools openapi-generator project.
It will probably be released soon.

https://github.com/OpenAPITools/openapi-generator/commit/0da409986889dcfab1420fb1410ae9509a5834c8

Are you aware of this?
Do you know what the differences or (dis-)/advantages are from your project?
Cheers!

**Comment by koxudaxi** [comment]
@oktavlachs 

I didn't know it.

> Do you know what the differences or (dis-)/advantages are from your project?

I guess `openapi-generator-cli` can support openapi features like multiple status codes of responses. 
But, We can implement the features to fastapi-code-generator.
And the code-generator creates project files (tests, docker file, ....)

However, `openapi-generator-cli` can generate only a simple model.
https://github.com/OpenAPITools/openapi-generator/commit/0da409986889dcfab1420fb1410ae9509a5834c8#diff-b667e564efa49b2fa6b6d766c52b062dc73f3e6444046e34d9457d6e532e91e0

`fastapi-code-generator` use `datamodel-code-generator` to generate models.
The `datamodel-code-generator` supports a lot of features. 
For example, `allOf` is created as an inherited class.
I believe definitely this is the best advantage.

Btw, `datamodel-code-generator` is used in a lot of projects.
It's download about [100k per month](https://pepy.tech/project/datamodel-code-generator)
Famous tech compies convert complex and large Schema files to pydantic models by `datamodel-code-generator` . 😉 
-  Netflix https://github.com/Netflix/consoleme
- Airbyte https://github.com/airbytehq/airbyte
- IBM https://github.com/IBM/compliance-trestle
- Seldon https://github.com/SeldonIO/MLServer
- Qwant https://github.com/Qwant/idunn

If fastapi-code-generator can call all features of datamodel-code-generator then I guess fastapi-code-generator becomes an awesome tool to generate fastapi and pydnatic code.

**Comment by oktavlachs** [comment]
@koxudaxi 

That makes sense and I get the idea that pydantic models will be far better to adjust also then the ones generated by the cli project. Alright then, I am excited too see what you can make possible within your project.

Can I somehow contribute?

**Comment by koxudaxi** [comment]
@oktavlachs 
Sure, I always welcome PR and suggestions.
If you have ideas for a new feature, then you can create an issue for each request.
Also, you can create PRs directly if the changes are small.
Thank you!!

**Comment by koxudaxi** [comment]
@oktavlachs 
I have released a new version `0.3.0` 
The version use parser of the `datamodel-code-generator`.

I tried to parse your schema.
Unfortunately, the code generator can't parse the schema yet.
I will check the problem of `datamodel-code-generator`.

**Comment by oktavlachs** [comment]
@koxudaxi 
Awesome, thanks for keeping this up. I am sure when that is solved there will be nothing stopping me from using your solution for code generation!
Just keep in mind that there might also be a bug in the spec, as they were discovered in other 3gpp OAS3 files in the past. But what you describe sounds more like some internal stuff in your project maybe.
Anyway, thanks and would be happy if you sent an update at some point :)
Cheers
